// Copyright 2021 MozoLM Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Server-side configuration.

syntax = "proto3";

package mozolm.grpc;

import "mozolm/models/model_config.proto";

option java_package = "com.google.mozolm.grpc";
option java_outer_classname = "ServerConfigProto";
option java_multiple_files = true;

enum CredentialType {
  CREDENTIAL_INSECURE = 0;
  CREDENTIAL_SSL = 1;
}

// Next available ID: 3
message ServerAuthConfig {
  // Type of authorization credentials used in client/server handshake.
  CredentialType credential_type = 1;

  // Next available ID: 5
  message SslConfig {
    // Private server key for SSL.
    string server_key = 1;

    // Public server certificate.
    string server_cert = 2;

    // Custom certificate authority.
    string custom_ca = 3;

    // Whether a valid client certificate is required.
    bool client_verify = 4;
  }

  // Configuration for a secure gRPC channel. If not given, insecure connection
  // is used.
  SslConfig ssl_config = 2;
}

// Next available ID: 6
message ServerConfig {
  // Model hub configuration.
  ModelHubConfig model_hub_config = 1;

  // The address to try to bind to the server in URI form. If
  // the scheme name is omitted, "dns:///" is assumed. To bind to any address,
  // please use IPv6 any, i.e., [::]:<port>, which also accepts IPv4
  // connections.  Valid values include `dns:///localhost:1234`,
  // `192.168.1.1:31416`, `dns:///[::1]:27182`, etc.
  //
  // To use Unix, rather than TCP, sockets use the "unix:///" scheme. Example:
  // `unix:///tmp/example.sock`.
  string address_uri = 2;

  // Configuration for authentication mechanisms.
  ServerAuthConfig auth = 3;

  // Whether the server will wait for client requests. Always enabled by
  // default, only disabled for the testing.
  bool wait_for_clients = 4;

  // Number of threads for handling requests asynchronously.
  int32 async_pool_size = 5;
}
